elki-outlier/src/main/java/elki/outlier/density/HySortOD.java

```diff
@@ -0,0 +1,743 @@
+        sorted.add(h);
+      }
+      else {
+        obj.add(objectPosition);
+      }
+
+      objectPosition++;
+    }
+
+    int n = sorted.size();
+    List<Hypercube> H = new ArrayList<Hypercube>(n);
+
+    this.invertedIndex = new int[relation.size()];
+    int hypercybePosition = 0;
+    for(Hypercube h : sorted) {
+      H.add(h);
+
+      for(Integer _objectPosition : h.getInstances())
+        this.invertedIndex[_objectPosition] = hypercybePosition;
+
+      hypercybePosition++;
+    }
+
+    return H;
+  }
+
+  private double score(int density) {
+    return 1 - (density / this.strategy.getMaxDensity());
+  }
+
+  @Override
+  public TypeInformation[] getInputTypeRestriction() {
+    return TypeUtil.array(TypeUtil.NUMBER_VECTOR_FIELD);
+  }
+
+  class Hypercube implements Comparable<Hypercube>, Comparator<Hypercube> {
+
+    int[] coords;
+
+    ArrayList<Integer> instances;
+
+    public Hypercube(V instance, double lenght) {
+      super();
+      this.coords = new int[instance.getDimensionality()];
+      this.instances = new ArrayList<Integer>();
+
+      for(int d = 0; d < instance.getDimensionality(); d++) {
+        this.coords[d] = (int) Math.floor(instance.doubleValue(d) / lenght);
+      }
+    }
+
+    @Override
+    public int compare(Hypercube o1, Hypercube o2) {
+      for(int i = 0; i < o1.coords.length; i++) {
+        int d = o1.coords[i] - o2.coords[i];
+        if(d != 0)
+          return d;
+      }
+      return 0;
+    }
+
+    @Override
+    public int compareTo(Hypercube o) {
+      for(int i = 0; i < coords.length; i++) {
+        int d = this.coords[i] - o.coords[i];
+        if(d != 0)
+          return d;
+      }
+      return 0;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+      if(Objects.isNull(obj))
+        return false;
+
+      if(getClass() != obj.getClass())
+        return false;
+
+      @SuppressWarnings("unchecked")
+      final Hypercube other = (Hypercube) obj;
+
+      if(coords.length != other.coords.length)
+        return false;
+
+      for(int i = 0; i < coords.length; i++)
+        if(coords[i] != other.coords[i])
+          return false;
+
+      return true;
+    }
+
+    @Override
+    public String toString() {
+      StringBuilder str = new StringBuilder();
+      str.append("(" + coords[0]);
+      for(int i = 1; i < coords.length; i++)
+        str.append(", " + coords[i]);
+      str.append(")");
+      return str.toString();
+    }
+
+    public int getCoordAt(int j) {
+      return coords[j];
+    }
+
+    public int[] getCoords() {
+      return coords;
+    }
+
+    public int getNumDimensions() {
+      return coords.length;
+    }
+
+    public List<Integer> getInstances() {
+      return instances;
+    }
+
+    public void add(Integer instance) {
+      instances.add(instance);
+    }
+
+    public int getDensity() {
+      return instances.size();
+    }
+
+  }
+
+  /**
+   * 
+   * @source https://www.baeldung.com/java-binary-tree
```

Comment:
Including source code copied from some web site is a licensing problem.
But for all I can tell, the BinaryTree may be pretty much a Java TreeSet (which is a red-black tree if I am not mistaken).